### PR TITLE
tensor == tensor should be bool

### PR DESCRIPTION
**Pull request by Qazalin**
this is [blocking dtypes math](https://github.com/tinygrad/tinygrad/pull/2800#issuecomment-1867796741) because argmax is the wrong dtype.
we already support NEG on bools:
```
kernel void E_4(device bool* data0, const device int* data1, const device int* data2, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
  int gidx0 = gid.x; /* 4 */
  int val0 = *(data1+gidx0);
  int val1 = *(data2+gidx0);
  *(data0+gidx0) = (!((val0<val1)+(val1<val0)));
}
```
@chenyuxyz lmk if CMPEQ is already in the workings tho

### CODE REVIEW

**Comment by chenyuxyz** [comment]
CMPEQ is blocked due to issues in llvm. if you fixed the tests (likely some places assumed it's float without casting) we can merge this first

**Comment by Qazalin** [comment]
i think we need COMPEQ if we dont wanna skip [this test](https://github.com/tinygrad/tinygrad/actions/runs/7304048409/job/19905607568?pr=2916#step:8:263) in Metal CI, for some reason `!((cast0<val3)+(val3<cast0))` returns the flipped result in the actions runners version.

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -819,10 +819,10 @@ def __ixor__(self, x) -> Tensor: return self.assign(self.xor(x))
   # in webgpu bool cannot be used as a storage buffer type
   def __lt__(self, x) -> Tensor: return mlops.Less.apply(*self._broadcasted(x, False)).cast(dtypes.float if self.device == "WEBGPU" else dtypes.bool)
   def __gt__(self, x) -> Tensor: return mlops.Less.apply(*self._broadcasted(x, True)).cast(dtypes.float if self.device == "WEBGPU" else dtypes.bool)
-  def __ge__(self, x) -> Tensor: return 1.0-(self<x)
-  def __le__(self, x) -> Tensor: return 1.0-(self>x)
+  def __ge__(self, x) -> Tensor: return -(self<x) if self.device != "WEBGPU" else 1.0-(self<x)
+  def __le__(self, x) -> Tensor: return -(self>x) if self.device != "WEBGPU" else 1.0-(self>x)
   def __ne__(self, x) -> Tensor: return (self<x) + (self>x)   # type: ignore[override]
```
Comment (chenyuxyz) [inline tinygrad/tensor.py:824]: does `((self<x).float() + (self>x).float()).cast(dtypes.bool)` work? i think it's fine for short-term workaround before CMPEQ

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -659,7 +659,7 @@ def fix(x:Tensor): return x.reshape(*ret.shape[0:-2], ret.shape[-2] * ret.shape[
   @staticmethod
   def _tri(r:sint, c:sint, k:int=0, **kwargs) -> Tensor:
     assert all_int((r,c)), "does not support symbolic"
-    return Tensor.arange(r, **kwargs).unsqueeze(1).expand(r,c) <= Tensor.arange(-k, c-k, **kwargs).unsqueeze(0).expand(r,c)
+    return (Tensor.arange(r, **kwargs).unsqueeze(1).expand(r,c) <= Tensor.arange(-k, c-k, **kwargs).unsqueeze(0).expand(r,c)).float()
```
Comment (chenyuxyz) [inline tinygrad/tensor.py:662]: this does not looks right, it's the input to where?

**Comment by Qazalin** [comment]
tbh I'd take the tests and fixes from here and just do CMPEQ, ended up not blocking dtypes math anyways.

File: tinygrad/tensor.py (hunk 2 above)
Comment (Qazalin) [reply tinygrad/tensor.py:662]: when you compare two aranges they get fused and the acc becomes bool, we need sum to be at least int32
https://gist.github.com/Qazalin/38316f9c3d06be9ecb0057519333e5ed

Comment (Qazalin) [reply tinygrad/tensor.py]: now it's correct
```
#include <metal_stdlib>
using namespace metal;
kernel void r_5_5(device bool* data0, const device int* data1, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
  int gidx0 = gid.x; /* 5 */
  int acc0 = 0;
  int val0 = *(data1+0);
  for (int ridx0 = 0; ridx0 < 5; ridx0++) {
    acc0 = (((((gidx0*(-1))+(ridx0*(-1)))<(-3))?1:0)+acc0);
  }
  *(data0+gidx0) = (!((acc0+(-2))<val0));
}
```

**Comment by Qazalin** [comment]
the compeq PR #2931 also did the .neg() changes, this is just the tests now

**Comment by chenyuxyz** [comment]
test looks good, can you sync to master?

**Pull request merged**